Conditional variants

becker33 opened the pull request:
A common question from users has been how to model variants that are new in new versions of a package, or variants that are dependent on other variants. Our stock answer so far has been an unsatisfying combination of "just have it do nothing in the old versoin" and "tell Spack it conflicts".

This PR enables conditional variants, on any spec condition. The syntax is straightforward, and matches that of previous features.

```
variant('version_based', default=False, when='@2.0:', description="Variant that is only available in versions 2.0 and later")
variant('variant_based', default=False, when='+version_based', description="Variant that depends on another variant")
```

PR includes tests.

Closes #9740
This is a prereq to addressing #14337

becker33 committed:
allow conditional variants

becker33 committed:
concretizer rules for conditional variants

becker33 committed:
add tests for conditional variants

adamjstewart commented:
Can't wait to test this on packages like boost and PyTorch! With the old concretizer, I have to disable half of the variants in PyTorch because they don't exist.

becker33 committed:
remove unreachable conditional

becker33 committed:
refactor for more pythonic representation

tgamblin commented:
@becker33: does this work well at all with the old concretizer?  I love this feature but had been thinking it would really suck without a proper solver. I guess we don't have to worry too much longer, as 0.17 will make the new one default.

becker33 committed:
make conditional variants fail when specified explicitly and ~condition

becker33 commented:
@tgamblin it seems to work reasonably well with the old concretizer -- it's at least passing all the tests there. I do think it increases the capability to express relationships that the old concretizer can't solve, but it won't break existing packages/variants until someone updates to use it, and at that point I think it's better to model reality well than conform to the old concretizer well, given that the new one will be default soon.

adamjstewart commented:
Tested this with `py-torch` with the following modifications:

<details>
<summary>Click to expand!</summary>

```diff
diff --git a/var/spack/repos/builtin/packages/py-torch/package.py b/var/spack/repos/builtin/packages/py-torch/package.py
index 63c7cac7cc..dc002ae210 100644
--- a/var/spack/repos/builtin/packages/py-torch/package.py
+++ b/var/spack/repos/builtin/packages/py-torch/package.py
@@ -49,51 +49,36 @@ class PyTorch(PythonPackage, CudaPackage):
 
     # All options are defined in CMakeLists.txt.
     # Some are listed in setup.py, but not all.
-    variant('caffe2', default=True, description='Build Caffe2')
+    variant('caffe2', default=True, when='@1.7:', description='Build Caffe2')
     variant('cuda', default=not is_darwin, description='Use CUDA')
-    variant('rocm', default=False, description='Use ROCm')
-    variant('cudnn', default=not is_darwin, description='Use cuDNN')
-    variant('fbgemm', default=True, description='Use FBGEMM (quantized 8-bit server operators)')
-    variant('kineto', default=True, description='Use Kineto profiling library')
-    variant('magma', default=not is_darwin, description='Use MAGMA')
+    variant('rocm', default=False, when='@1.0:', description='Use ROCm')
+    variant('cudnn', default=not is_darwin, when='@1.0: +cuda', description='Use cuDNN')
+    variant('fbgemm', default=True, when='@1.0:', description='Use FBGEMM (quantized 8-bit server operators)')
+    variant('kineto', default=True, when='@1.8:', description='Use Kineto profiling library')
+    variant('magma', default=not is_darwin, when='+cuda', description='Use MAGMA')
     variant('metal', default=is_darwin, description='Use Metal for Caffe2 iOS build')
-    variant('nccl', default=not is_darwin, description='Use NCCL')
+    variant('nccl', default=not is_darwin, when='+cuda platform=linux', description='Use NCCL')
+    variant('nccl', default=not is_darwin, when='+cuda platform=cray', description='Use NCCL')
+    variant('nccl', default=not is_darwin, when='+rocm platform=linux', description='Use NCCL')
+    variant('nccl', default=not is_darwin, when='+rocm platform=cray', description='Use NCCL')
     variant('nnpack', default=True, description='Use NNPACK')
-    variant('numa', default=not is_darwin, description='Use NUMA')
+    variant('numa', default=not is_darwin, when='platform=linux', description='Use NUMA')
+    variant('numa', default=not is_darwin, when='platform=cray', description='Use NUMA')
     variant('numpy', default=True, description='Use NumPy')
     variant('openmp', default=True, description='Use OpenMP for parallel code')
-    variant('qnnpack', default=True, description='Use QNNPACK (quantized 8-bit operators)')
-    variant('valgrind', default=not is_darwin, description='Use Valgrind')
-    variant('xnnpack', default=True, description='Use XNNPACK')
-    variant('mkldnn', default=True, description='Use MKLDNN')
+    variant('qnnpack', default=True, when='@1.0:', description='Use QNNPACK (quantized 8-bit operators)')
+    variant('valgrind', default=not is_darwin, when='@1.8: platform=linux', description='Use Valgrind')
+    variant('valgrind', default=not is_darwin, when='@1.8: platform=cray', description='Use Valgrind')
+    variant('xnnpack', default=True, when='@1.5:', description='Use XNNPACK')
+    variant('mkldnn', default=True, when='@1.0:', description='Use MKLDNN')
     variant('distributed', default=not is_darwin, description='Use distributed')
-    variant('mpi', default=not is_darwin, description='Use MPI for Caffe2')
-    variant('gloo', default=not is_darwin, description='Use Gloo')
-    variant('tensorpipe', default=not is_darwin, description='Use TensorPipe')
-    variant('onnx_ml', default=True, description='Enable traditional ONNX ML API')
-
-    conflicts('+cuda', when='+rocm')
-    conflicts('+cudnn', when='~cuda')
-    conflicts('+magma', when='~cuda')
-    conflicts('+nccl', when='~cuda~rocm')
-    conflicts('+nccl', when='platform=darwin')
-    conflicts('+numa', when='platform=darwin', msg='Only available on Linux')
-    conflicts('+valgrind', when='platform=darwin', msg='Only available on Linux')
-    conflicts('+mpi', when='~distributed')
-    conflicts('+gloo', when='~distributed')
-    conflicts('+tensorpipe', when='~distributed')
-    conflicts('+kineto', when='@:1.7.999')
-    conflicts('+valgrind', when='@:1.7.999')
-    conflicts('~caffe2', when='@0.4.0:1.6.999')  # no way to disable caffe2?
-    conflicts('+caffe2', when='@:0.3.1')  # caffe2 did not yet exist?
-    conflicts('+tensorpipe', when='@:1.5.999')
-    conflicts('+xnnpack', when='@:1.4.999')
-    conflicts('~onnx_ml', when='@:1.4.999')  # no way to disable ONNX?
-    conflicts('+rocm', when='@:0.4.999')
-    conflicts('+cudnn', when='@:0.4.999')
-    conflicts('+fbgemm', when='@:0.4.999,1.4.0')
-    conflicts('+qnnpack', when='@:0.4.999')
-    conflicts('+mkldnn', when='@:0.4.999')
+    variant('mpi', default=not is_darwin, when='+mpi', description='Use MPI for Caffe2')
+    variant('gloo', default=not is_darwin, when='+mpi', description='Use Gloo')
+    variant('tensorpipe', default=not is_darwin, when='@1.6: +mpi', description='Use TensorPipe')
+    variant('onnx_ml', default=True, when='@1.5:', description='Enable traditional ONNX ML API')
+
+    conflicts('+cuda+rocm')
+    conflicts('+fbgemm', when='@1.4.0')
 
     conflicts('cuda_arch=none', when='+cuda',
               msg='Must specify CUDA compute capabilities of your GPU, see '
```
</details>

Here is what I observed:

1. This works perfectly with both the original and clingo concretizers!
1. With `depends_on`, I can simulate "or" by using multiple lines. However, the following doesn't seem to work:
   ```python
    variant('nccl', default=not is_darwin, when='+cuda platform=linux', description='Use NCCL')
    variant('nccl', default=not is_darwin, when='+cuda platform=cray', description='Use NCCL')
    variant('nccl', default=not is_darwin, when='+rocm platform=linux', description='Use NCCL')
    variant('nccl', default=not is_darwin, when='+rocm platform=cray', description='Use NCCL')
    ```
1. Another common feature request is variant defaults that change based on other variants. I don't think this is possible with the current implementation, right? I'm not sure what this would even look like...

Another idea that would be interesting. Right now there are a lot of packages with something like:
```python
if self.spec.satisfies('@1.2:'):
    if '+foo' in self.spec:
        args.append('--enable-foo')
    else:
        args.append('--disable-foo')
```
GDAL is a good example of this. I'm wondering if the following would also work if the `+foo` variant only exists in 1.2+:
```python
if '+foo' in self.spec:
    args.append('--enable-foo')
elif '~foo' in self.spec:
    args.append('--disable-foo')
```

becker33 commented:
@adamjstewart thanks for checking this out!

This PR extends the when syntax to allow a list of conditions (for `or`). If you use that syntax, you can get `or` to work properly instead of using multiple lines.

```
variant('nccl', default=not is_darwin, when='+cuda platform=linux', description='Use NCCL')
variant('nccl', default=not is_darwin, when='+cuda platform=cray', description='Use NCCL')
variant('nccl', default=not is_darwin, when='+rocm platform=linux', description='Use NCCL')
variant('nccl', default=not is_darwin, when='+rocm platform=cray', description='Use NCCL')
```
could become
```
variant('nccl', default=not is_darwin, when=['+cuda platform=linux', '+cuda platform=cray', '+rocm platform=linux', '+rocm platform=cray'], description='Use NCCL')
```
The reason I did that is because separate defaults per-declaration cannot work without some really substantial work, and I don't want people to try
```
variant('foo', default=False, when='@1.1')
variant('foo', default=True, when='@1.2:')
```

However, as long as only the when-clause is changing, I can probably make that work in the interest of not confusing people who are used to the old syntax. (The new syntax does work for all the other commands that accept when clauses though, so it should catch on well enough for "or"). I'll take a look at making the duplicate variant declarations work for `or`.

adamjstewart commented:
> This PR extends the when syntax to allow a list of conditions

Does this work for all directives, not just variant? That's really cool!

becker33 committed:
Allow multiply defined variants for multiple when clauses

This is in addition to allowing multiple whens per-definition
Multiply defined variants must have the same default/values/validators

becker33 commented:
Yep, it works for all directives! I don't have tests in for all of them, but I have tests in for `depends_on` and all of them except for `variant` use the same code to handle it.

Also, multiply defined variants work now as long as you don't try to change the defaults/values/validators.

alalazo commented on lib/spack/spack/build_systems/autotools.py:
```diff
@@ -440,7 +440,8 @@ def _activate_or_not(
         # option and whether or not that option is activated
-        if set(self.variants[name].values) == set((True, False)):
+        variant_desc, _ = self.variants[name]
+        if set(variant_desc.values) == set((True, False)):
```
If I am reading this correctly here we are treating the absence of a variant as if it was set to `False` and so we'll emit e.g. a `--disable-foo` if `self.enable_or_disabled('foo')` is called. Just wanted to double check this is the semantics we want, rather than just disregarding it.

becker33 replied:
I think in this case it would throw an exception on line 443, because it unconditionally gets the variant from a dictionary.

alalazo commented on lib/spack/spack/cmd/info.py:
```diff
@@ -67,36 +67,41 @@ def __init__(self, variants):
             max(cols, 70) - max_name - max_vals - 14,
         )
-        self.column_widths = (max_name, max_vals, max_description)
+        self.column_widths = (max_name, max_when, max_vals, max_description)
```
This is a minor point, but we should probably revisit formatting. Right now I have this on my screen:
![Screenshot from 2021-07-27 16-38-19](https://user-images.githubusercontent.com/4199709/127174175-89ba7973-a424-477a-bb06-71a655027ddc.png)

which is ok but may be improved in terms of:
1. Wrapping the Description line
2. Joining the when clauses for variants and avoid `[]` when it is always present

adamjstewart replied:
It would be nice to not display the `when` column at all if it isn't used by a package.

alalazo replied:
The output looks better to me, thanks! There are still a few super-minor issues - for instance this is from `spack info gromacs`:
![Screenshot from 2021-11-02 19-51-46](https://user-images.githubusercontent.com/4199709/139927237-cdd8bd17-1ca6-43e0-9bec-867b1b3e0048.png)

I think though that we shouldn't hold this PR on that and we may fix minor issues like this in v0.17.1

alalazo commented on lib/spack/spack/directives.py:
```diff
@@ -62,6 +62,17 @@ class OpenMpi(Package):
 _patch_order_index = 0
 
 
+def make_when_specs(value):
```
The function name is very similar to `make_when_spec` so maybe we should change it. Also it would probably be good to prepend both names with an underscore to make it clear they are used only in this module.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -613,8 +613,11 @@ def pkg_rules(self, pkg, tests):
+        for name, entry in sorted(pkg.variants.items()):
+            variant, when = entry
+            for w in when:
+                cond_id = self.condition(w, name=pkg.name)
+                self.gen.fact(fn.variant_condition(pkg.name, name, cond_id))
```
Other `*_condition` facts:
```suggestion
                cond_id = self.condition(w, name=pkg.name)
                self.gen.fact(fn.variant_condition(cond_id, pkg.name, name))
```
put the ID first.

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -345,6 +345,15 @@ external_conditions_hold(Package, LocalIndex) :-
 %-----------------------------------------------------------------------------
+% a variant is a variant of a package if it is a variant under some condition
+% and that condition holds
+variant(Package, Variant) :- variant_condition(Package, Variant, ID),
```
```suggestion
variant(Package, Variant) :- variant_condition(ID, Package, Variant),
```

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -345,6 +345,15 @@ external_conditions_hold(Package, LocalIndex) :-
+variant(Package, Variant) :- variant_condition(Package, Variant, ID),
+                             condition_holds(ID).
+
+% a variant cannot be set if it is not a variant on the package
+:- variant_set(Package, Variant),
+   not variant(Package, Variant).
```
I'll poke with this more, but we may also need an integrity constraint to prohibit a variant to take some value if the variant condition is not met:
```
:- variant_value(Package, Variant, _), not variant(Package, Variant).
```

alalazo reviewed (changes requested):
Generally this looks great. I tried modifying `cp2k` and I can make all the CUDA related variants disappear if `~cuda` is in the spec. This should give a sensible improvement in readability for complex specs!

The only major thing: I have been a bit puzzled by the extension of the `when=` argument to accept a list of specs. My first impression is that it is not obvious that this would be an OR and it is effectively syntactic sugar for having two directives with different `when=` arguments. It also deviates a bit from the declarative nature of directives. I guess I have two questions:

- Is this feature technically necessary to get conditional variants?
- Should we at least check what would be user reactions to it (mine and @adamjstewart seem to be opposite)?


Also, this slight modification:
```diff
diff --git a/lib/spack/spack/directives.py b/lib/spack/spack/directives.py
index 45d9fd52d3..9b21e99bb3 100644
--- a/lib/spack/spack/directives.py
+++ b/lib/spack/spack/directives.py
@@ -252,7 +252,7 @@ def _wrapper(*args, **kwargs):
                 if DirectiveMeta._when_constraints_from_context:
                     # Check that directives not yet supporting the when= argument
                     # are not used inside the context manager
-                    if decorated_function.__name__ in ('version', 'variant'):
+                    if decorated_function.__name__ in ('version',):
                         msg = ('directive "{0}" cannot be used within a "when"'
                                ' context since it does not support a "when=" '
                                'argument')
```
is needed to make this feature work with the `when` context manager.

alalazo commented:
@becker33 Would it be possible to split this in two PRs:
1. Allow conditional variants
2. Make `when=` arguments accept a list of conditions

and get 1. merged soon?

becker33 committed:
concretizer: add constraint no variant value for variant not on package

becker33 committed:
reorder variant_condition args to match other condition types

becker33 committed:
spack info: improve conditional variant format

becker33 committed:
fix variant directive docstring

haampie commented:
This is a such a cool feature to get in! It greatly reduces the verbosity of specs with many variants, it solves the long standing comment in cuda.py:

```
    # FIXME: keep cuda and cuda_arch separate to make usage easier until
    # Spack has depends_on(cuda, when='cuda_arch!=None') or alike
```

and the same for rocm archs. It removes 50 characters from a spec that uses both rocm and cuda :D.

I tend to agree with @alalazo's comment that the new *or* syntax for `when` is orthogonal to the main contribution of this PR. Also note that we now have the `with when(...)` syntax, which does not support the same list-means-OR concept? If you would not add that it would be inconsistent, if you would add it, it would lead to potentially complex package definitions that requires a first-order logic refresher course to mentally reduce the following to their disjunctive normal form 

```python
with when(['+a', '+b']):
  depends_on('x', when=['+c', '+d'])
```

which is `depends_on('x', when=['+a +c', '+b +c', '+a +d', '+b +d'])`.

Maybe we can think about that more after 0.17?

alalazo commented:
Are there cross-references between this PR and #26719 i.e. do we need to add new messages to #26719 if this PR is merged first?

becker33 commented:
@alalazo yes, we will need to add one message

alalazo commented:
I think conditional variants will be most effective when applied to mixins like `CudaPackage` or `RocmPackage`(in the sense that a lot of packages will benefit from them if we change those base classes). Should we try to get these two base classes modified for v0.17.0 (in this PR or in a following one)? Asking about that since it's a hash changing modification so if we miss v0.17.0 and stick to our own rules we won't see it in a release until v0.18.0.

becker33 committed:
fixup

becker33 committed:
valid hashes for mock package

becker33 committed:
add error messages for new clingo failure modes

becker33 committed:
allow variant directive in when context manager

alalazo commented on lib/spack/spack/directives.py:
```diff
@@ -640,14 +643,30 @@ def _raise_default_not_set(pkg):
+        variant_desc = spack.variant.Variant(name, default, description,
+                                             values, multi, validator)
+
+        if name in pkg.variants:
+            orig_variant, orig_when = pkg.variants[name]
+            if variant_desc != orig_variant:
+                msg = ("multiply defined variant %s " % name +
+                       "has incompatible declarations. Only when-clauses " +
+                       "may differ between variant declarations.")
+                raise DirectiveError(msg)
```
I added the following to `cp2k`:
```python
variant('foo', default=True, description='', when='+mpi')
variant('foo', default=False, description='', when='+openmp')
```
but I couldn't trigger this error.

alalazo replied:
Can we remove this error check completely? With the comparison bug solved, this will not permit overriding a variant from a base package:
```console
$ spack solve cp2k
==> Error: cannot bootstrap the "clingo" Python module from spec "clingo-bootstrap@spack+python %gcc target=x86_64" due to the following failures:
    'github-actions' raised DirectiveError: multiply defined variant build_type has incompatible declarations. Only when-clauses may differ between variant declarations.
    'spack-install' raised DirectiveError: multiply defined variant build_type has incompatible declarations. Only when-clauses may differ between variant declarations.
    Please run `spack -d spec zlib` for more verbose error messages
```

alalazo replied:
Solved in ca5eb36

alalazo commented on lib/spack/spack/variant.py:
```diff
@@ -181,6 +181,17 @@ def variant_cls(self):
+                self.values == other.values and
+                self.multi == other.multi and
+                self.single_value_validator == other.single_value_validator and
+                self.group_validator, other.group_validator)
```
I think there's a bug here:
```suggestion
                self.group_validator == other.group_validator)
```
Unfortunately this will make a lot of other tests fail due to the check mentioned [in this comment]( https://github.com/spack/spack/pull/24858/files#r741370209)

alalazo replied:
Solved in ca5eb36

becker33 committed:
address review bug

alalazo commented:
The failure on Fedora is happening on many PRs. It is unrelated to changes and possibly due to synchronization issues with upstream when new Fedora releases are pushed.

adamjstewart commented:
Did we ever add support for `when=[...]`? It seems like that feature was removed from this PR.
